File: 39-combination-sum/combination-sum.py
class Solution:

    def combinationSum(self, arr: List[int], target: int) -> List[List[int]]:
        ans = []

        def help(arr, target, curr, i, lst):
            if i >= len(arr) or curr > target:
                return
            if curr == target:
                ans.append(lst[:])
                return

            lst.append(arr[i])
            help(arr, target, curr + arr[i], i, lst)

            lst.pop()
            help(arr, target, curr, i + 1, lst)
        help(arr, target, 0, 0, [])
        return ans

File: 39-combination-sum/combination-sum.java
class Solution {
    List<List<Integer>> ans = new ArrayList<>();

    public void help(int arr[], int target, int curr, int i, List<Integer> lst) {
        if (curr > target || i >= arr.length)
            return;
        if (curr == target) {
            ans.add(new ArrayList<>(lst));
            return;
        }
        lst.add(arr[i]);
        help(arr, target, curr + arr[i], i, lst);

        lst.remove(lst.size() - 1);
        help(arr, target, curr, i + 1, lst);
    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        help(candidates, target, 0, 0, new ArrayList<>());
        return ans;
    }
}